parlai/tasks/redial/agents.py

```diff
@@ -0,0 +1,137 @@
+#!/usr/bin/env python3
+
+# Copyright (c) Facebook, Inc. and its affiliates.
+# This source code is licensed under the MIT license found in the
+# LICENSE file in the root directory of this source tree.
+
+from parlai.core.teachers import FixedDialogTeacher
+from .build import build
+import os
+import json
+import re
+import csv
+
+
+def _path(opt):
+    # build the data if it does not exist
+    build(opt)
+
+    # set up path to data (specific to each dataset)
+    jsonl_dirpath = os.path.join(opt['datapath'], 'redial')
+    return jsonl_dirpath
+
+
+# Turns title from format "Title (Year)" to "Title" or leaves as is if no (Year)
+def remove_year_from_title(title):
+    matches = re.finditer(r"\s\(", title)
+    indices = [m.start(0) for m in matches]
+    if indices:
+        title_end = indices[-1]
+        return title[:title_end]
+    else:
+        return title
+
+
+def replace_movie_ids(id_string, id_map):
+    pattern = r'@\d+'
+    return re.sub(pattern, lambda s: id_map[s.group()], id_string)
+
+
+class ReDialTeacher(FixedDialogTeacher):
+    """
+    ReDial Teacher.
+    """
+
+    def __init__(self, opt, shared=None):
+        super().__init__(opt, shared)
+        jsonl_path = _path(opt)
+        self.title_id_map = {}
+        self.get_title_dict(jsonl_path)
+        if shared is not None:
+            self.episodes = shared['episodes']
+        else:
+            self.episodes = []
+            self._setup_data(jsonl_path)
+        self.id = 'redial'
+
+        self.reset()
+
+    def get_title_dict(self, path):
+        csv_path = os.path.join(path, 'movies_with_mentions.csv')
+        with open(csv_path, mode='r') as f:
+            reader = csv.reader(f)
+            for row in reader:
+                self.title_id_map['@' + row[0]] = remove_year_from_title(row[1])
+
+    def _setup_data(self, jsonl_path):
+        train_path = os.path.join(jsonl_path, 'train_data.jsonl')
+        test_path = os.path.join(jsonl_path, 'test_data.jsonl')
+        valid_split = 0.5
```

Comment:
where does this number come from? could you perhaps leave a comment? (i.e. if this is specified in the paper)

Reply:
They split the data set up 80/10/10, so I assumed the test data was split 50/50. Looking at this now this would be closer to 90/5/5, so I think it might be better to get valid data  from the train data and leave the test data as is, since that would be closer to 80/10/10.